### ISSUE TITLE
pmacct does to work on RPi4/CM4

### ISSUE THREAD

**Issue by MigNov** [pmacct/pmacct#595]
**Description**
Hello,
I've been trying to compile pmacct project on the RPi4 and RPi CM4 but it seems to be not supported as it's returning the "Bus error". According to several sources, including https://forums.raspberrypi.com/viewtopic.php?f=33&t=102400, it seems to have the issue with hard-floating ABI but when I try to set the CFLAGS to "-mcpu=cortex-a72 -mfloat-abi=soft -mfpu=crypto-neon-fp-armv8" it breaks the compilation entirely. Is this project going to support ARM/RPi4?

Thanks a lot for your answer,
Michal

**Version**
$ nfacctd -V
NetFlow Accounting Daemon, nfacctd 1.7.8-git [20220409-3 (d09284b0)]

Arguments:
 '--enable-l2' '--enable-traffic-bins' '--enable-bgp-bins' '--enable-bmp-bins' '--enable-st-bins'

Libs:
cdada 0.4.0
libpcap version 1.10.0 (with TPACKET_V3)

Plugins:
memory
print
nfprobe
sfprobe
tee

System:
Linux 5.15.30-v8+ #1536 SMP PREEMPT Mon Mar 28 13:53:14 BST 2022 aarch64

Compiler:
gcc 10.2.1

**Comment by MigNov** [comment]
Hi Paolo (@paololucente ),

thanks for the information. I have verified the situation and it really seems to be caused by the nfprobe plugin as it's working smoothly with `print` and `memory` plugin so it is really related to the `nfprobe` plugin and sender. At least I know where to look now. Unfortunately I don't have the RPi 4 nor RPi CM4 available on any public IP address so I cannot provide you with the access. Were you also the author of `nfprobe` plugin or did somebody else implement the `nfprobe` plugin ?

Thanks a lot,
Michal

**Comment by paololucente** [comment]
Hi Michal,

Great, we have two consistent experiences. `nfprobe` plugin was forked from `softflowd` several years ago (perhaps even around 10 years ago); while i did not write the original code, i'd say i am familiar enough with it and its differences from the original code base.

I would then say, please send over your config and an output of the crash from gdb so to nail where in the code the program is failing. Should you not be familiar with gdb, please see here: https://github.com/pmacct/pmacct/blob/master/QUICKSTART#L2915-L2935

Paolo

**Comment by MigNov** [comment]
Hi Paolo,
I've tried to debug myself and it seems to be somewhere in the process_packet() function of the `nfprobe` plugin but I was unable to pin-point the source of the issue. The core dump is available here: [https://mig-cz.cf/core-pmacct.core](url) [I cannot upload core dump file directly]. System is Raspbian 11 Bullseye with gcc 10 and "Linux  5.15.30-v8+ #1536 SMP PREEMPT Mon Mar 28 13:53:14 BST 2022 aarch64 GNU/Linux".

The configuration file is:
```
daemonize: false
aggregate: src_host, dst_host, src_port, dst_port, proto, tos
plugins: nfprobe
nfprobe_receiver: 172.29.10.100:2055
nfprobe_version: 10
nfprobe_timeouts: tcp=30:maxlife=60
```
and the command line I'm using to run it is:

`pmacctd -i eth0 -f ../pmacctd.conf`

Thanks,
Michal

**Comment by paololucente** [comment]
Hi Michal,

Given the difference in architecture, ie. i don't have access to a Pi4, the core dump file itself is slightly less useful. Can you copy/paste the output of a `bt` done on the gdb console? ie.:

* `gdb /path/to/pmacctd /path/to/coredump file` then
* `gdb>` prompt type `bt` and
* copy/paste the output in this Issue

Paolo

**Comment by paololucente** [comment]
Hi Michal,

Unfortunately it turns this does not help much - it is very strange because the crash happens in a section of code that is also shared with other plugins, ie. `print`, but there the problem does not manifest itself. Without access to a RPi4/CM4, i think there is little i can do to progress this issue.

Without resolving the problem, trying to circumvent it, one last thing i may suggest is to try to get away from home-grown internal queueing (default) and switch to ZeroMQ queueing. You can read a bit more about it here: https://github.com/pmacct/pmacct/blob/master/QUICKSTART#L1172-L1230 . You could compile pmacct with ZeroMQ support and add / adapt the following lines to your config: https://github.com/pmacct/pmacct/blob/master/QUICKSTART#L1224-L1226 to enable the feature.

Paolo

**Comment by MigNov** [comment]
Hi Paolo,
thanks a lot for your time and your information. I've been digging into that a little deeper and I also did the test on my Raspberry Pi 4 (not Compute Module 4) I have at home and it passed. According to the official information both RPi 4 and RPi CM4 should be having the same chipset but the strange thing is that it was working in one case and in the latter case it was not working at all resulting into the SIGBUS signal. After some time I have realized that the issue is with the architecture of the system as I'm using 32-bit system on my home RPi 4 but on the testing RPi CM4 I have AArch64 so this looks like the issue with AArch64 support for libpcap or other library.

Thanks for the information regarding switch to ZeroMQ enablement, I will try to use it as well.

Thanks a lot,
Michal
